### PR TITLE
Fingerprinting

### PR DESCRIPTION
**Pull request by ptpaterson**
Ticket(s): [FE-3011](https://faunadb.atlassian.net/browse/FE-3011)

## Problem
We don't send the driver environment details.

## Solution
Send them!

## Result
Gathers and formats information to send via the `X-Driver-Env` header.

## Testing
Test added to make sure the header exists.
Integration tests will make sure this works in other environments besides Node.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

[FE-3011]: https://faunadb.atlassian.net/browse/FE-3011?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -6,7 +6,8 @@
     "outDir": "./dist",
     "declaration": true,
     "declarationDir": "./dist",
-    "esModuleInterop": true
+    "esModuleInterop": true,
+    "resolveJsonModule": true
```
Comment (cleve-fauna) [inline tsconfig.json:10]: why is this added?

File: src/util/environment.ts [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+  } catch (_) {
+    // okay if this fails
+  }
+
+  if (isNode()) {
+    driverEnv.runtime = ["nodejs", process.version].join("-");
+    driverEnv.env = getNodeRuntimeEnv();
+    driverEnv.os = [os.platform(), os.release()].join("-");
+  } else if (isServiceWorker) {
+    driverEnv.runtime = "Service Worker";
```
Comment (pnwpedro) [inline src/util/environment.ts:51]: Which cases does this condition cover, and are we sure that os won't be available/useful, or would it be better to try to obtain them?

Comment (ptpaterson) [reply src/util/environment.ts:51]: I don't think the executing context (global scope) for a service worker has anything else useful in it. I think it's just the [ServiceWorkerGlobalScope](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope).

Comment (ptpaterson) [reply src/util/environment.ts:51]: I see [ServiceWorkerGlobalScope.navigator](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/navigator), though. We should be able to get the browser info from that.

I'll work on that 👍🏻

Comment (ptpaterson) [reply src/util/environment.ts]: Cool! updated so ServiceWorker should send the same fields as a browser.

File: src/client.ts [hunk 3]
```diff
@@ -439,6 +440,8 @@ in an environmental variable named FAUNA_SECRET or pass it to the Client\
     ) {
       headerObject["x-last-txn-ts"] = this.#lastTxnTs;
     }
+
+    headerObject["X-Driver-Env"] = getDriverEnv();
```
Comment (cleve-fauna) [inline src/client.ts:444]: let's cache this once at construction time so we don't have to farm this on every request.

File: src/util/environment.ts [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+  if (typeof process !== "undefined" && process.release?.name === "node") {
+    try {
+      require("node:http2");
+      return true;
+    } catch (_) {
+      return false;
+    }
+  }
+  return false;
+};
```
Comment (cleve-fauna) [inline src/util/environment.ts:24]: this does not tell us if its node - but if its node with HTTP2.

We should use slightly different algorithms for fingerprinting vs determining a default client.

This one can be used for the default client - and can stay where it was (perhaps with a different name like useNodeHttp2) and a new one can just determine if its node or not.

File: src/util/environment.ts [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+/**
+ * Get OS details for the browser
+ */
+const getBrowserOsDetails = (
+  navigator: Navigator | WorkerNavigator
+): string => {
+  let os = "unknown";
+  const clientStrings = [
+    { s: "Windows 10", r: /(Windows 10.0|Windows NT 10.0)/ },
```
Comment (cleve-fauna) [inline src/util/environment.ts:161]: windows 11 is the latest windoes and is not present in this list

File: package.json [hunk 6]
```diff
@@ -20,6 +20,7 @@
     "@tsconfig/node18": "^1.0.1",
     "@types/jest": "^29.4.0",
     "@types/node": "^18.13.0",
+    "@types/serviceworker": "^0.0.67",
```
Comment (cleve-fauna) [inline package.json:23]: We don't seem to actually use this package.

**Review by cleve-fauna — COMMENTED** [review]
Some suggestions.

File: package.json (hunk 6 above)
Comment (ptpaterson) [reply package.json:23]: This one is necessary for Typescript to understand the `ServiceWorkerGlobalScope` and `WorkerNavigator` types.

File: src/util/environment.ts (hunk 5 above)
Comment (ptpaterson) [reply src/util/environment.ts:161]: good catch! I am looking into this and my understanding is now that it is not possible to distinguish between Windows 10 and 11 through the user agent string. Chromium is freezing user agents strings in favor of newerm methods, which I think means the [User-Agent Client Hints API](https://developer.mozilla.org/en-US/docs/Web/API/User-Agent_Client_Hints_API). Unfortunately though, it's still in an experimental state and at least firefox doesn't support said methods yet.

https://developers.whatismybrowser.com/learn/browser-detection/user-agents/detect-windows-11-user-agent

I suggest we proceed as is, and consider following up with something like:

1. check if the user-agent hints API is available and use that
2. fall back on current impl

File: package.json (hunk 6 above)
Comment (cleve-fauna) [reply package.json:23]: where do we use those types? I must have overlooked them. Mind linking me?

Comment (cleve-fauna) [reply package.json:23]: nm i found one

**Review by cleve-fauna — COMMENTED** [review]
Two minor suggestions.

**Review by cleve-fauna — APPROVED** [review]
